src/native/corehost/browserhost/browserhost.cpp

```diff
@@ -0,0 +1,173 @@
+        unsigned int domainId,
+        int* latchedExitCode);
+
+    pal::hresult_t coreclr_execute_assembly(
+        void* hostHandle,
+        unsigned int domainId,
+        int argc,
+        const char** argv,
+        const char* managedAssemblyPath,
+        unsigned int* exitCode);
+
+    pal::hresult_t coreclr_create_delegate(
+        void* hostHandle,
+        unsigned int domainId,
+        const char* entryPointAssemblyName,
+        const char* entryPointTypeName,
+        const char* entryPointMethodName,
+        void** delegate);
+
+    pal::hresult_t coreclr_set_error_writer(
+        coreclr_error_writer_callback_fn error_writer);
+
+    const void* SystemResolveDllImport(const char* name);
+    const void* SystemJSResolveDllImport(const char* name);
+    const void* SystemJSInteropResolveDllImport(const char* name);
+    const void* GlobalizationResolveDllImport(const char* name);
+    const void* CompressionResolveDllImport(const char* name);
+
+    bool browserHostExternalAssemblyProbe(const char* pathPtr, /*out*/ void **outDataStartPtr, /*out*/ int64_t* outSize);
+    void browserHostResolveMain(int exitCode);
+    void browserHostRejectMain(const char *reason);
+}
+
+// The current CoreCLR instance details.
+static void* CurrentClrInstance;
+static unsigned int CurrentAppDomainId;
+
+static void log_error_info(const char* line)
+{
+    std::fprintf(stderr, "log error: %s\n", line);
+}
+
+static bool external_assembly_probe(const char* path, /*out*/ void **data_start, /*out*/ int64_t* size)
+{
+    *size = 0;
+    *data_start = nullptr;
+    return browserHostExternalAssemblyProbe(path, data_start, size);;
+}
+
+static const void* pinvoke_override(const char* library_name, const char* entry_point_name)
+{
+    if (strcmp(library_name, "libSystem.Native") == 0)
+    {
+        return SystemResolveDllImport(entry_point_name);
+    }
+    if (strcmp(library_name, "libSystem.JavaScript") == 0)
+    {
+        return SystemJSResolveDllImport(entry_point_name);
+    }
+    if (strcmp(library_name, "libSystem.Runtime.InteropServices.JavaScript") == 0)
+    {
+        return SystemJSInteropResolveDllImport(entry_point_name);
+    }
+    if (strcmp(library_name, "libSystem.Globalization.Native") == 0)
+    {
+        return GlobalizationResolveDllImport(entry_point_name);
+    }
+    if (strcmp(library_name, "libSystem.IO.Compression.Native") == 0)
+    {
+        return CompressionResolveDllImport(entry_point_name);
+    }
+
+    return nullptr;
+}
+
+static pal::string_t app_path;
+static pal::string_t search_paths;
+static pal::string_t tpa;
+static const pal::string_t app_domain_name = "corehost";
+static const pal::string_t exe_path = "/managed";
+static std::vector<const char*> propertyKeys;
+static std::vector<const char*> propertyValues;
+static pal::char_t buffer[STRING_LENGTH("0xffffffffffffffff")];
+
+extern "C" int browserHostInitializeCoreCLR(void)
+{
+    //WASM-TODO: does getenv return UTF8 ?
```

Comment:
It does on emscripten wasi but i can't find the environ implementation for browser wasm anywhere in emscripten's repo